tfe/resource_tfe_team_members_test.go

```diff
@@ -200,10 +226,12 @@ resource "tfe_team" "foobar" {
 
 resource "tfe_team_members" "foobar" {
   team_id   = "${tfe_team.foobar.id}"
-  usernames = ["admin", "sander"]
-}`
+  usernames = ["admin", "%s"]
+}`, username)
+}
 
-const testAccTFETeamMembers_update = `
+func testAccTFETeamMembers_update(username string) string {
```

Comment:
I think making these functions which both take a `username` makes it a bit less clear what is being tested. Any reason why you prefer this format over following the same approach I used here: https://github.com/terraform-providers/terraform-provider-tfe/blob/master/tfe/resource_tfe_oauth_client_test.go#L106-L118

So use global variables (e.g. `TFE_USER1`, `TFE_USER2`) defined once, versus redefining them in every test that needs it? As I could see a use for these same vars in these tests as well: https://github.com/terraform-providers/terraform-provider-tfe/blob/master/tfe/resource_tfe_team_member_test.go#L12-L77

Reply:
> I think making these functions which both take a `username` makes it a bit less clear what is being tested.

You're right. I'm thinking about removing `testAccTFETeamMembers_update` completely and changing the interface of the first function to this

```go
func testAccTFETeamMembers_basic(usernames []string) string
```

that way it's more obvious from the body of the test that we're changing one username as the subject of test (list of usernames) is more explicit and visible. Also we can reduce LOC that way as the same function can be used for both stages of the test (creation & update).

> Any reason why you prefer this format over following the same approach I used here

So the "function with variables"-approach is used throughout the AWS provider and many other major providers. The main reason being that it can scale (the same config can be reused anywhere) and provides compile-time validation (as opposed to plain `fmt.Sprintf` which would just render `%(missing variable)` missing variables or `%(incorrect type)`for wrong order/type).

Lastly passing a variable to function call makes it more obvious what's being tested in the place where it's being tested (test body). i.e. I don't need to look at the whole template to understand that a username is (a) even configurable and (b) where does it come from.

That said I'd be open to adding a helper function to the SDK (at some point) so we can source such variables without having to import `os`, but I'm not sure if there's much benefit, other than swapping one import for another. We'd probably need to figure out how to couple it more tightly with `testAccCheckEnvVariables` (which should also be native part of the SDK).

> As I could see a use for these same vars in these tests as well

That's fine because these tests don't seem to be calling an API, so it's safe to keep your username hardcoded there as we're not constrained by its real existence in TFE IMO?

Reply:
> You're right. I'm thinking about removing `testAccTFETeamMembers_update` completely and changing the interface of the first function to this

I think that would make it even less clear to be honest 😂 And I think you wouldn't need a slice of usernames in there right? The body only uses a single username variable.

As for the rest of this discussion, I'm fine with changing it to a function. I'm just (overly) sensitive to inconsistencies. So currently this provider only uses `consts` or `vars` that contain the configs. Introducing this new approach versus reusing the existing approach makes me shiffer a bit and makes me want to change all the tests inputs to functions so its consistent again 😉

Reply:
> And I think you wouldn't need a slice of usernames in there right? The body only uses a single username variable.

Well I'd make all usernames configurable, i.e. I'd pull out `admin` too.

> I'm fine with changing it to a function. I'm just (overly) sensitive to inconsistencies.

So I'm happy to retain consistency within the TFE provider and change other `const`s to `func`s, but I just learned to live in a chaotic inconsistent world where things are moving, mainly because maintaining consistency effectively is hard in all the big providers. 😂

Reply:
> but I just learned to live in a chaotic inconsistent world

Guess I'm not quite there yet.. Still trying to hold on 😂